conformance: `TestGatewayConformance/HTTPRouteWeight/Requests_should_have_a_distribution_that_matches_the_weight` is flaky in CI

### Problem statement

`TestGatewayConformance/HTTPRouteWeight/Requests_should_have_a_distribution_that_matches_the_weight` is flaky in CI

```
=== RUN   TestGatewayConformance/HTTPRouteWeight
    conformance.go:72: 2025-09-05T08:35:13.166313977Z: Applying tests/httproute-weight.yaml
    apply.go:275: 2025-09-05T08:35:13.217526157Z: Creating weighted-backends HTTPRoute
    conformance.go:77: 2025-09-05T08:35:13.22315634Z: Running HTTPRouteWeight, relying on the following features: Gateway-standard, HTTPRoute-standard
    helpers.go:611: 2025-09-05T08:35:13.245180504Z: Conditions matched expectations
    helpers.go:611: 2025-09-05T08:35:13.245269691Z: Route gateway-conformance-infra/weighted-backends Parents matched expectations
=== RUN   TestGatewayConformance/HTTPRouteWeight/Requests_should_have_a_distribution_that_matches_the_weight
    httproute-weight.go:68: 2025-09-05T08:35:13.254900845Z: Making GET request to http://172.18.128.1/
    http.go:251: 2025-09-05T08:35:13.259575663Z: Request passed
    httproute-weight.go:94: 2025-09-05T08:35:13.259632199Z: Making GET request to http://172.18.128.1/
    httproute-weight.go:72: Traffic distribution test failed (1/10): backend "infra-backend-v1" weighted traffic of 1 not within tolerance 0.7 (+/-0.050000)
        expect traffic to hit backend "infra-backend-v2" - but none was received
        expected only two backends to receive traffic
    httproute-weight.go:94: 2025-09-05T08:35:13.557421215Z: Making GET request to http://172.18.128.1/
    httproute-weight.go:72: Traffic distribution test failed (2/10): backend "infra-backend-v1" weighted traffic of 1 not within tolerance 0.7 (+/-0.050000)
        expect traffic to hit backend "infra-backend-v2" - but none was received
        expected only two backends to receive traffic
    httproute-weight.go:94: 2025-09-05T08:35:13.821619019Z: Making GET request to http://172.18.128.1/
    httproute-weight.go:72: Traffic distribution test failed (3/10): backend "infra-backend-v1" weighted traffic of 1 not within tolerance 0.7 (+/-0.050000)
        expect traffic to hit backend "infra-backend-v2" - but none was received
        expected only two backends to receive traffic
    httproute-weight.go:94: 2025-09-05T08:35:14.115073971Z: Making GET request to http://172.18.128.1/
    httproute-weight.go:72: Traffic distribution test failed (4/10): backend "infra-backend-v1" weighted traffic of 1 not within tolerance 0.7 (+/-0.050000)
        expect traffic to hit backend "infra-backend-v2" - but none was received
        expected only two backends to receive traffic
    httproute-weight.go:94: 2025-09-05T08:35:14.427694733Z: Making GET request to http://172.18.128.1/
    httproute-weight.go:72: Traffic distribution test failed (5/10): backend "infra-backend-v1" weighted traffic of 1 not within tolerance 0.7 (+/-0.050000)
        expect traffic to hit backend "infra-backend-v2" - but none was received
        expected only two backends to receive traffic
    httproute-weight.go:94: 2025-09-05T08:35:14.703737188Z: Making GET request to http://172.18.128.1/
    httproute-weight.go:72: Traffic distribution test failed (6/10): backend "infra-backend-v1" weighted traffic of 1 not within tolerance 0.7 (+/-0.050000)
        expect traffic to hit backend "infra-backend-v2" - but none was received
        expected only two backends to receive traffic
    httproute-weight.go:94: 2025-09-05T08:35:14.973636398Z: Making GET request to http://172.18.128.1/
    httproute-weight.go:72: Traffic distribution test failed (7/10): backend "infra-backend-v1" weighted traffic of 1 not within tolerance 0.7 (+/-0.050000)
        expect traffic to hit backend "infra-backend-v2" - but none was received
        expected only two backends to receive traffic
    httproute-weight.go:94: 2025-09-05T08:35:15.273781023Z: Making GET request to http://172.18.128.1/
    httproute-weight.go:72: Traffic distribution test failed (8/10): backend "infra-backend-v1" weighted traffic of 1 not within tolerance 0.7 (+/-0.050000)
        expect traffic to hit backend "infra-backend-v2" - but none was received
        expected only two backends to receive traffic
    httproute-weight.go:94: 2025-09-05T08:35:15.526611522Z: Making GET request to http://172.18.128.1/
    httproute-weight.go:72: Traffic distribution test failed (9/10): backend "infra-backend-v1" weighted traffic of 1 not within tolerance 0.7 (+/-0.050000)
        expect traffic to hit backend "infra-backend-v2" - but none was received
        expected only two backends to receive traffic
    httproute-weight.go:94: 2025-09-05T08:35:15.808083173Z: Making GET request to http://172.18.128.1/
    httproute-weight.go:72: Traffic distribution test failed (10/10): backend "infra-backend-v1" weighted traffic of 0.806 not within tolerance 0.7 (+/-0.050000)
        backend "infra-backend-v2" weighted traffic of 0.194 not within tolerance 0.3 (+/-0.050000)
    httproute-weight.go:77: Weighted distribution tests failed
--- FAIL: TestGatewayConformance/HTTPRouteWeight/Requests_should_have_a_distribution_that_matches_the_weight (2.98s)
```

https://github.com/Kong/kong-operator/actions/runs/17487917526/job/49670864031?pr=2214#step:7:2538

--- a/controller/gateway/controller_consts.go
+++ b/controller/gateway/controller_consts.go
@@ -1,5 +1,9 @@
 package gateway
 
+import (
+	"github.com/kong/kong-operator/v2/pkg/consts"
+)
+
 // -----------------------------------------------------------------------------
 // Gateway - Finalizers
 // -----------------------------------------------------------------------------
@@ -14,14 +18,11 @@ const (
 	GatewayFinalizerCleanupControlPlanes GatewayFinalizer = "gateway-operator.konghq.com/cleanup-controlplanes"
 	// GatewayFinalizerCleanupNetworkPolicies is the finalizer to cleanup owned network policies.
 	GatewayFinalizerCleanupNetworkPolicies GatewayFinalizer = "gateway-operator.konghq.com/cleanup-network-policies"
-	// KonnectGatewayControlPlaneFinalizer is the finalizer added to KonnectGatewayControlPlane resources.
 )
 
 // -----------------------------------------------------------------------------
 // KonnectGatewayControlPlane - Finalizers
 // -----------------------------------------------------------------------------
 
-const (
-	// KonnectGatewayControlPlaneFinalizer is the finalizer added to KonnectGatewayControlPlane resources when created by the Gateway controller.
-	KonnectGatewayControlPlaneFinalizer = "gateway.operator.konghq.com/konnect-gateway-controlplane"
-)
+// KonnectGatewayControlPlaneFinalizer is the finalizer added to KonnectGatewayControlPlane resources when created by the Gateway controller.
+const KonnectGatewayControlPlaneFinalizer = consts.KonnectGatewayControlPlaneFinalizer

--- a/controller/konnect/reconciler_generic.go
+++ b/controller/konnect/reconciler_generic.go
@@ -507,16 +507,25 @@ func (r *KonnectEntityReconciler[T, TEnt]) Reconcile(
 				}
 				return ctrl.Result{}, err
 			}
-			if err := r.Client.Update(ctx, ent); err != nil {
-				if apierrors.IsConflict(err) {
-					return ctrl.Result{Requeue: true}, nil
-				}
-				// in case the finalizer removal fails because the resource does not exist, ignore the error.
-				if apierrors.IsNotFound(err) {
-					return ctrl.Result{}, nil
-				}
-				return ctrl.Result{}, fmt.Errorf("failed to remove finalizer %s: %w", KonnectCleanupFinalizer, err)
+		}
+
+		// For KonnectGatewayControlPlane resources, also remove the finalizer added
+		// by the Gateway controller. This handles the race condition where a Gateway
+		// is deleted before the Konnect reconciler adds its own finalizer, leaving
+		// only the Gateway controller's finalizer on the KonnectGatewayControlPlane
+		// with no controller able to remove it (since the parent Gateway is gone).
+		if _, ok := any(ent).(*konnectv1alpha2.KonnectGatewayControlPlane); ok {
+			controllerutil.RemoveFinalizer(ent, consts.KonnectGatewayControlPlaneFinalizer)
+		}
+
+		if err := r.Client.Update(ctx, ent); err != nil {
+			if apierrors.IsConflict(err) {
+				return ctrl.Result{Requeue: true}, nil
 			}
+			if apierrors.IsNotFound(err) {
+				return ctrl.Result{}, nil
+			}
+			return ctrl.Result{}, fmt.Errorf("failed to update resource after finalizer removal: %w", err)
 		}
 
 		return ctrl.Result{}, nil

--- a/pkg/consts/konnect.go
+++ b/pkg/consts/konnect.go
@@ -14,6 +14,10 @@ const (
 // -----------------------------------------------------------------------------
 
 const (
+	// KonnectGatewayControlPlaneFinalizer is the finalizer added to KonnectGatewayControlPlane
+	// resources when created by the Gateway controller.
+	KonnectGatewayControlPlaneFinalizer = "gateway.operator.konghq.com/konnect-gateway-controlplane"
+
 	// CleanupPluginBindingFinalizer is the finalizer that is attached to entities that
 	// are referenced as targets by managed KongPluginBindings (binding instances created
 	// by the controller out of entities' konghq.com/plugins annotations).

--- a/test/conformance/skipped_tests_test.go
+++ b/test/conformance/skipped_tests_test.go
@@ -7,10 +7,6 @@ import (
 )
 
 var skippedTestsShared = []string{
-	// TODO: https://github.com/Kong/kong-operator/issues/2215
-	// HTTPRouteWeight is flaky for some reason. 2215 tracks solving it.
-	tests.HTTPRouteWeight.ShortName,
-
 	// NOTE:
 	// Issue tracking all gRPC related conformance tests:
 	// https://github.com/Kong/kong-operator/issues/2345

